### PR TITLE
Add secure redis

### PR DESCRIPTION
**Pull request by andresrinivasan**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
Redis can only be used without authentication

Issue Number: #2449 

## What is the new behavior?
Redis can be used with authentication

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Are there any specific instructions or things that should be known prior to reviewing?

This is tightly coupled to 
    [developer-scripts PR#248](https://github.com/edgexfoundry/developer-scripts/pull/248)
    [go-mod-bootstrap PR#68](https://github.com/edgexfoundry/go-mod-bootstrap/pull/68)

## Other information

### CODE REVIEW

**Comment by andresrinivasan** [comment]
I have no idea why the PR is showing I changed 97 files. I changed/added

```
.gitignore
Makefile
SECURITY.md
cmd/core-command/res/configuration.toml
cmd/core-data/res/configuration.toml
cmd/core-metadata/res/configuration.toml
cmd/security-file-token-provider/res/token-config.json
cmd/security-secrets-setup/README.md
cmd/security-secretstore-read/Attribution.txt
cmd/security-secretstore-read/README.md
cmd/security-secretstore-read/main.go
cmd/security-secretstore-read/res/configuration.toml
cmd/security-secretstore-setup/Dockerfile
cmd/security-secretstore-setup/README.md
cmd/security-secretstore-setup/entrypoint.sh
cmd/support-logging/res/configuration.toml
cmd/support-notifications/res/configuration.toml
cmd/support-scheduler/res/configuration.toml
go.mod
internal/pkg/bootstrap/handlers/database/database.go
internal/security/secretread/config.go
internal/security/secretread/constants.go
internal/security/secretread/init.go
internal/security/secretstore/init.go
```

File: cmd/security-secretstore-read/main.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"os"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	"github.com/edgexfoundry/go-mod-core-contracts/models"
+
+	"github.com/edgexfoundry/edgex-go/internal/security/secretread"
+)
+
+func main() {
+	lc := logger.NewClient(secretread.ServiceKey, false, "logs/security-secretstore-read.log", models.ErrorLog)
```
Comment (bnevis-i) [inline cmd/security-secretstore-read/main.go:28]: @lenny-intel: Lenny, any comments on what we should be doing here WRT logging?

@andresrinivasan: Don't really like the relative logging path hard coded in.  Will be harder to make the container root FS read-only later.

File: cmd/security-secretstore-setup/entrypoint.sh [hunk 2]
```diff
@@ -47,10 +47,15 @@ echo "Initializing secret store"
...
+# Share Redis v5 password
+mkdir -p /secrets
+chown -Rh 100:1000 /secrets
+/security-secretstore-read -confdir /res-read >${REDIS5_PASSWORD_PATHNAME}
+
 # write a sentinel file when we're done because consul is not
 # secure and we don't trust it it access to the EdgeX secret store
 if [ ! -z "${SECRETSTORE_SETUP_DONE_FLAG}" ]; then
-    echo "Signaling secretstore-setup completion"
+    echo "Signaling secretstore-setup completion"
```
Comment (bnevis-i) [inline cmd/security-secretstore-setup/entrypoint.sh:58]: Surious diff.

File: cmd/support-logging/res/configuration.toml [hunk 3]
```diff
@@ -24,13 +24,7 @@ File = ''
...
 [Databases]
   [Databases.Primary]
-  Host = 'localhost'
-  Name = 'logging'
-  Password = 'password'
-  Port = 27017
-  Username = 'logging'
-  Timeout = 5000
-  Type = 'mongodb'
+  Type = 'file'
```
Comment (bnevis-i) [inline cmd/support-logging/res/configuration.toml:27]: Why this change?

File: cmd/security-secretstore-setup/entrypoint.sh (hunk 2 above)
Comment (andresrinivasan) [reply cmd/security-secretstore-setup/entrypoint.sh:58]: Type-o. Will correct.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -4,6 +4,7 @@ Please check if your PR fulfills the following requirements:
 - [ ] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
 - [ ] Tests for the changes have been added (for bug fixes / features)
 - [ ] Docs have been added / updated (for bug fixes / features)
+- [ ] `make test` has completed successfully
```
Comment (bnevis-i) [inline .github/PULL_REQUEST_TEMPLATE.md:7]: This probably should have been a separate PR.

File: cmd/security-secretstore-read/main.go (hunk 1 above)
Comment (andresrinivasan) [reply cmd/security-secretstore-read/main.go:28]: Agreed this is weak sauce. I am committed to properly integrate with go-mod-bootstrap but I ran out of time as I was getting this to work in time for code freeze.

File: cmd/support-logging/res/configuration.toml (hunk 3 above)
Comment (andresrinivasan) [reply cmd/support-logging/res/configuration.toml:27]: Logger needs to default to a file when Redis is the default database.

File: .gitignore [hunk 5]
```diff
@@ -27,6 +27,7 @@ cmd/support-scheduler/support-scheduler
 cmd/sys-mgmt-agent/sys-mgmt-agent
 cmd/sys-mgmt-executor/sys-mgmt-executor
 cmd/security-file-token-provider/security-file-token-provider
+cmd/security-secretstore-read/security-secretstore-read
```
Comment (bnevis-i) [inline .gitignore:30]: :heavy_check_mark:

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 4 above)
Comment (andresrinivasan) [reply .github/PULL_REQUEST_TEMPLATE.md:7]: Reverting and creating a separate PR.

File: Makefile [hunk 6]
```diff
@@ -64,6 +70,9 @@ cmd/security-secretstore-setup/security-secretstore-setup:
 cmd/security-file-token-provider/security-file-token-provider:
 	$(GO) build $(GOFLAGS) -o ./cmd/security-file-token-provider/security-file-token-provider ./cmd/security-file-token-provider
 
+cmd/security-secretstore-read/security-secretstore-read:
+	$(GO) build $(GOFLAGS) -o $@ ./cmd/security-secretstore-read
+
```
Comment (bnevis-i) [inline Makefile:75]: :heavy_check_mark:

File: cmd/security-file-token-provider/res/token-config.json [hunk 7]
```diff
@@ -127,6 +176,15 @@
...
           ]
+        },
+        "secret/edgex/redis/*": {
+          "capabilities": [
+            "list",
+            "read",
+            "create",
+            "update",
+            "delete"
+          ]
```
Comment (bnevis-i) [inline cmd/security-file-token-provider/res/token-config.json:187]: :heavy_check_mark: (whole file)

File: cmd/security-secretstore-setup/Dockerfile [hunk 8]
```diff
@@ -29,9 +31,12 @@ WORKDIR /
 COPY --from=builder /edgex-go/cmd/security-file-token-provider/res/token-config.json /res-file-token-provider/token-config.json
 COPY --from=builder /edgex-go/cmd/security-secretstore-setup/res-file-token-provider/configuration.toml  /res-file-token-provider/configuration.toml
 COPY --from=builder /edgex-go/cmd/security-secretstore-setup/res/configuration.toml /res/configuration.toml
+COPY --from=builder /edgex-go/cmd/security-secretstore-read/res/configuration.toml /res-read/configuration.toml
+
```
Comment (bnevis-i) [inline cmd/security-secretstore-setup/Dockerfile:35]: :heavy_check_mark: (whole file)

File: go.mod [hunk 9]
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/OneOfOne/xxhash v1.2.5
 	github.com/cloudflare/gokey v0.1.0
 	github.com/dgrijalva/jwt-go v3.2.0+incompatible
-	github.com/edgexfoundry/go-mod-bootstrap v0.0.27
+	github.com/edgexfoundry/go-mod-bootstrap v0.0.29
```
Comment (bnevis-i) [inline go.mod:9]: :heavy_check_mark:

File: cmd/support-logging/res/configuration.toml (hunk 3 above)
Comment (bnevis-i) [reply cmd/support-logging/res/configuration.toml:27]: Ok, but there is no condition here.  This is changed for all use cases.

File: cmd/security-secretstore-read/main.go (hunk 1 above)
Comment (bnevis-i) [reply cmd/security-secretstore-read/main.go:28]: Noted the lack of integration with go-mod-bootstrap.  Will not make further comments on it in this PR.

File: SECURITY.md [hunk 10]
```diff
@@ -3,33 +3,37 @@
 Starting with the Fuji release, EdgeX includes enhanced security features that are enabled by default.
 There are 3 major components that are responsible for security features:
 
-* [Security-secrets-setup](cmd/security-secrets-setup/README.md)
-* [Security-secretstore-setup](cmd/security-secretstore-setup/README.md)
-* [Security-proxy-setup](cmd/security-proxy-setup/README.md)
+| Component  | Description  |
```
Comment (bnevis-i) [inline SECURITY.md:6]: Nice additions to the document.

File: cmd/security-secrets-setup/README.md [hunk 11]
```diff
@@ -95,7 +99,7 @@ To see what the TLS materials are generated inside the docker container,
 one can run the built docker image above (assuming the version number is `1.1.0`) as:
 
 ```sh
-$ docker run --rm -it edgexfoundry/docker-edgex-vault:1.1.0-dev /bin/sh
+docker run --rm -it edgexfoundry/docker-edgex-vault:1.1.0-dev /bin/sh
```
Comment (bnevis-i) [inline cmd/security-secrets-setup/README.md:102]: Whoops... This is out of date, or the container name is wrong.  We are using a stock "vault" image now instead of a custom one.  Probably needs to be updated to docker-edgex-secrets-setup-go:master-dev

File: cmd/security-secretstore-setup/README.md [hunk 12]
```diff
@@ -39,3 +39,42 @@ make docker_security_secretstore_setup
...
+* To explore the vault
+
+    ```sh
+    docker exec -ti edgex-vault sh -l
+    export VAULT_SKIP_VERIFY=true
+    export VAULT_TOKEN=s.xxxxxxxxxxxxxxxx
+    vault kv list secret/
+    ```
+
+    and drill down from there. To read a key use `vault read`.
```
Comment (bnevis-i) [inline cmd/security-secretstore-setup/README.md:73]: `vault kv get`

File: cmd/security-secretstore-setup/entrypoint.sh [hunk 13]
```diff
@@ -47,6 +47,11 @@ echo "Initializing secret store"
 echo "Executing custom command: $@"
 "$@"
 
+# Share Redis v5 password
+mkdir -p /secrets
+chown -Rh 100:1000 /secrets
+/security-secretstore-read -confdir /res-read >${REDIS5_PASSWORD_PATHNAME}
+
```
Comment (bnevis-i) [inline cmd/security-secretstore-setup/entrypoint.sh:54]: This is buggy.  Presuming from the above, that the secret store is the Record of Origin for the redis password, we shouldn't be taking it out of the secret store and putting it in a persistent docker volume.  This means that `REDIS5_PASSWORD_PATHNAME` should probably be of the form /tmp/edgex/secrets/edgex-redis/redis5-password  instead of /secrets/redis5-password.  Moreover, we should be guarding this with an "if" statement and not execute this block of code if `REDIS5_PASSWORD_PATHNAME` is not defined.  Also we should not assume the directory is /secrets but should be computing `basename ${REDIS5_PASSWORD_PATHNAME}` and chmodding from there.

File: cmd/security-secretstore-read/res/configuration.toml [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+#################################################################################
+
+[Writable]
+  LogLevel = "ERROR"
+
+[Service]
+  BootTimeout = 30000
+
+[SecretStore]
+  Host = "edgex-vault"
```
Comment (bnevis-i) [inline cmd/security-secretstore-read/res/configuration.toml:23]: @tonyespy Do you like to see docker hostname or `localhost` in the default configuration.toml?

**Comment by bnevis-i** [comment]
Done reviewing for now.  Will wait for up-to-date docker-compose and do a system-level test next.

File: cmd/security-secretstore-read/main.go (hunk 1 above)
Comment (andresrinivasan) [reply cmd/security-secretstore-read/main.go:28]: Tracking go-mod-bootstrap issue via #2503

File: cmd/security-secrets-setup/README.md (hunk 11 above)
Comment (andresrinivasan) [reply cmd/security-secrets-setup/README.md:102]: @bnevis-i Can I suggest a separate issue for this one? All I did was cleanup the command line so it was an easy copy/paste.

Comment (bnevis-i) [reply cmd/security-secrets-setup/README.md:102]: Sure.

File: cmd/security-secretstore-setup/entrypoint.sh (hunk 13 above)
Comment (andresrinivasan) [reply cmd/security-secretstore-setup/entrypoint.sh:54]: WIP

File: cmd/security-secretstore-setup/Dockerfile [hunk 15]
```diff
@@ -14,7 +14,9 @@ RUN go mod download
 
 COPY . .
 
-RUN make cmd/security-file-token-provider/security-file-token-provider cmd/security-secretstore-setup/security-secretstore-setup
+RUN make cmd/security-file-token-provider/security-file-token-provider \
+  cmd/security-secretstore-setup/security-secretstore-setup\
```
Comment (jim-wang-intel) [inline cmd/security-secretstore-setup/Dockerfile:18]: Nit pick: for consistence sake, may leave empty space at the end before \

File: internal/security/secretread/config.go [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ * @author: Diana Atanasova
+ * @author: Andre Srinivasan
+ *******************************************************************************/
+package secretread
```
Comment (jim-wang-intel) [inline internal/security/secretread/config.go:17]: Just a question: Should this one also change to integrate with go-mod-bootstrap via other PR eventually?

Comment (bnevis-i) [reply internal/security/secretread/config.go:17]: https://github.com/edgexfoundry/edgex-go/issues/2503

File: cmd/security-secretstore-setup/README.md (hunk 12 above)
Comment (andresrinivasan) [reply cmd/security-secretstore-setup/README.md:73]:
```
edgex-vault:/# vault kv get /secret/edgex/redis/redis5
====== Data ======
Key         Value
---         -----
password    |sL58?wrNiTehO}.
username    redis5
edgex-vault:/# vault read /secret/edgex/redis/redis5
Key                 Value
---                 -----
refresh_interval    168h
password            |sL58?wrNiTehO}.
username            redis5
```

Will update to say `vault kv get` or `vault read`.

File: cmd/security-secretstore-setup/Dockerfile (hunk 15 above)
Comment (andresrinivasan) [reply cmd/security-secretstore-setup/Dockerfile]: Updated and push'ed

File: cmd/security-secretstore-setup/entrypoint.sh (hunk 13 above)
Comment (andresrinivasan) [reply cmd/security-secretstore-setup/entrypoint.sh]: Done

**Comment by bnevis-i** [comment]
@tingyuz  Would like some help testing.... Pull down this branch, run a make docker, then pull down the redis security enable docker-compose files, change the :master images to :master-dev (for the ones built by edgex-go) and give it a shot.

**Comment by tingyuz** [comment]
I am having some trouble to bring the core services successfully. After running the modified docker-compose file I noticed the core services ( coredata/matedata/command etc) were running a bit then just quit. Some timing related issues here.

**Comment by andresrinivasan** [comment]
@tingyuz Please share a log from one of the microservices that is quiting.

**Comment by tingyuz** [comment]
> @tingyuz Please share a log from one of the microservices that is quiting.

My mistake - I haven't updated the consul images for weeks which caused the failure. Just updated to the latest version and it is working as expected.

**Comment by bnevis-i** [comment]
@andresrinivasan  Merging is blocked.

```
[2020-04-22T19:01:52.567Z] go: extracting golang.org/x/text v0.3.2
[2020-04-22T19:01:53.654Z] ./bin/test-attribution-txt.sh
[2020-04-22T19:01:56.414Z] An attribution for github.com/go-redis/redis/v7 is missing from in cmd/security-secretstore-read Attribution.txt, please add
[2020-04-22T19:01:58.456Z] make: *** [test] Error 1
```

File: cmd/security-secretstore-read/Attribution.txt [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+gopkg.in/yaml.v2 (Apache 2.0) https://github.com/go-yaml/yaml/
+https://github.com/go-yaml/yaml/blob/v2.2.2/LICENSE
+
+cloudflare/gokey (BSD-3) https://github.com/cloudflare/gokey
+https://github.com/cloudflare/gokey/blob/master/LICENSE
+
+golang.org/x/crypto (Unspecified) https://github.com/golang/crypto
+https://github.com/golang/crypto/blob/master/LICENSE
+
+go-redis/redis (BSD-2) https://github.com/go-redis/redis/
```
Comment (bnevis-i) [inline cmd/security-secretstore-read/Attribution.txt:162]:
```suggestion
go-redis/redis (BSD-2) https://github.com/go-redis/redis/
github.com/go-redis/redis/v7 (BSD-2) https://github.com/go-redis/redis
```
